server/executor/transaction_runner.go

```diff
@@ -0,0 +1,160 @@
+package executor
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/kubeshop/tracetest/server/config"
+	"github.com/kubeshop/tracetest/server/model"
+)
+
+type TransactionRunner interface {
+	Run(context.Context, model.Transaction, model.RunMetadata, model.Environment) model.TransactionRun
+}
+
+func NewTransactionRunner(runner Runner, db model.Repository, config config.Config) PersistentTransactionRunner {
+	return PersistentTransactionRunner{
+		testRunner:          runner,
+		db:                  db,
+		executionChannel:    make(chan transactionRunJob, 1),
+		exit:                make(chan bool),
+		checkTestStateDelay: config.PoolingRetryDelay() / 2,
+	}
+}
+
+type transactionRunJob struct {
+	ctx context.Context
+	run model.TransactionRun
+}
+
+type PersistentTransactionRunner struct {
+	testRunner          Runner
+	db                  model.Repository
+	checkTestStateDelay time.Duration
+	executionChannel    chan transactionRunJob
+	exit                chan bool
+}
+
+func (r PersistentTransactionRunner) Run(ctx context.Context, transaction model.Transaction, metadata model.RunMetadata, environment model.Environment) model.TransactionRun {
+	run := model.NewTransactionRun(transaction)
+	run.Metadata = metadata
+	run.Environment = environment
+
+	run, _ = r.db.CreateTransactionRun(ctx, run)
+
+	r.executionChannel <- transactionRunJob{
+		ctx: ctx,
+		run: run,
+	}
+
+	return run
+}
+
+func (r PersistentTransactionRunner) Stop() {
+	r.exit <- true
+}
+
+func (r PersistentTransactionRunner) Start(workers int) {
+	for i := 0; i < workers; i++ {
+		go func() {
+			fmt.Println("PersistentTransactionRunner start goroutine")
+			for {
+				select {
+				case <-r.exit:
+					fmt.Println("PersistentTransactionRunner exit goroutine")
+					return
+				case job := <-r.executionChannel:
+					err := r.runTransaction(job.ctx, job.run)
+					if err != nil {
+						panic(err)
+					}
+				}
+			}
+		}()
+	}
+}
+
+func (r PersistentTransactionRunner) runTransaction(ctx context.Context, run model.TransactionRun) error {
+	run.State = model.TransactionRunStateExecuting
+	environment := run.Environment
+
+	var err error
+
+	for i := range run.Steps {
+		run, err = r.runTransactionStep(ctx, run, i, environment)
+		if err != nil {
+			return fmt.Errorf("could not execute step %d of transaction %s: %w", i, run.TransactionID, err)
+		}
+
+		if run.State == model.TransactionRunStateExecuting {
```

Comment:
Why do you need this condition? I imagine it's to prevent updating the env on error. In that case, I'd use an exit guard, like 
```
if run.state == error {
  break
}
environment = r.patchEnvironment(environment, run) 
```

The current implementation might actually lead to a bug, or unexpected behaviour, if the last step returns a state  other than `exectuing`, where the environments wouldn't be correctly replaced. Might not be a case now, but might happen later